src/reference/00-Getting-Started/07A-Task-Graph.md

```diff
@@ -0,0 +1,324 @@
+Once those targets are created using the last pattern matching rule,
+only then the system command is executed to link `main.o` and `hello.o` to `hello`.
+
+If you're just running `make`, you can focus on what you want as the target,
+and the exact timing and commands necessary to build the intermediate products are figured out by Make.
+We can think of this as dependency-oriented programming, or flow-based programming.
+Make is actually considered a hybrid system because while the DSL describes the task dependencies, the actions are delegated to system commands.
+
+#### Rake
+
+This hybridity is continued for Make successors such as Ant, Rake, and sbt.
+Take a look at the basic syntax for Rakefile:
+
+```ruby
+task name: [:prereq1, :prereq2] do |t|
+  # actions (may reference prereq as t.name etc)
+end
+```
+
+The breakthrough made with Rake was that it used a programming language to
+describe the actions instead of the system commands.
+
+#### Benefits of hybrid flow-based programming
+
+Rather than thinking `settings` as a key-value map,
+a better analogy would be to think of it as a DAG (directed acyclic graph)
+of tasks where the edges denote *happens-before*. Let's call this the _task graph_.
+There are several motivation to organizing the build this way.
+
+First is de-duplication. With flow-based programming, a task is executed only once even when it is depended by multiple tasks.
+For example, even when multiple tasks along the task graph depend on `compile in Compile`,
+the compilation will be executed exactly once.
+
+Second is parallel processing. Using the task graph, the task engine can
+schedule mutually non-dependent tasks in parallel.
+
+Third is the separation of concern and the flexibility.
+The task graph lets the build user wire the tasks together in different ways,
+while sbt and plugins can provide various features such as compilation and
+library dependency management as functions that can be reused.
+
+### Declaring dependency to other tasks
+
+In `build.sbt` DSL, we use `.value` method to express the dependency to
+another task or setting. The value method is special and may only be
+called in the argument to `:=` (or, `+=` or `++=`, which we'll see later).
+
+As a first example, consider defining the `scalacOption` that depends on
+`update` and `clean` tasks. Here are the definitions of these keys (from [Keys](../sxr/sbt/Keys.scala.html)):
+
+```scala
+val scalacOptions = taskKey[Seq[String]]("Options for the Scala compiler.")
+val update = taskKey[UpdateReport]("Resolves and optionally retrieves dependencies, producing a report.")
+val clean = taskKey[Unit]("Deletes files produced by the build, such as generated sources, compiled classes, and task caches.")
+```
+
+Here's how we can define `scalacOptions`:
+
+```scala
+scalacOptions := {
+  val ur = update.value
+  val x = clean.value
+  ur.allConfigurations.take(3)
+}
+```
+
+**Note**: The values calculated above are nonsensical for `scalaOptions`,
```

Comment:
This feels like a really weird thing to do. At least for me, on seeing an example like this, my instinct is to try and understand what it's trying to achieve, which would be pointless in this case. It would be a while before I noticed this note.